src/main/java/io/cdap/plugin/kudu/KuduSourceConfig.java

```diff
@@ -16,67 +16,152 @@
-  public String optTableName;
+  private String optTableName;
 
-  @Name("schema")
+  @Name(SCHEMA)
   @Description("Output schema from Kudu source")
-  public String optSchema;
+  private String optSchema;
 
   // Optional Fields
-
-  @Name("columns")
+  @Name(COLUMNS)
   @Description("Columns to be projected from the table")
   @Nullable
-  public String optColumnProjection;
+  private String optColumnProjection;
 
-  @Name("opt-timeout")
+  @Name(OPT_TIMEOUT)
   @Description("Specifies the user operation timeout in milliseconds.")
   @Nullable
-  public String optOperationTimeout;
+  private String optOperationTimeout;
+
+  public KuduSourceConfig(String referenceName, String optMasterAddresses, String optTableName, String optSchema,
+                          @Nullable String optColumnProjection, @Nullable String optOperationTimeout) {
+    this.referenceName = referenceName;
+    this.optMasterAddresses = optMasterAddresses;
+    this.optTableName = optTableName;
+    this.optSchema = optSchema;
+    this.optColumnProjection = optColumnProjection;
+    this.optOperationTimeout = optOperationTimeout;
+  }
+
+  private KuduSourceConfig(Builder builder) {
+    referenceName = builder.referenceName;
+    optMasterAddresses = builder.optMasterAddresses;
+    optTableName = builder.optTableName;
+    optSchema = builder.optSchema;
+    optColumnProjection = builder.optColumnProjection;
+    optOperationTimeout = builder.optOperationTimeout;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static Builder builder(KuduSourceConfig copy) {
+    return builder()
+      .setReferenceName(copy.referenceName)
+      .setOptMasterAddresses(copy.optMasterAddresses)
+      .setOptTableName(copy.optTableName)
+      .setOptSchema(copy.optSchema)
+      .setOptColumnProjection(copy.optColumnProjection)
+      .setOptOperationTimeout(copy.optOperationTimeout);
+  }
+
+  public String getReferenceName() {
+    return referenceName;
+  }
+
+  public String getOptMasterAddresses() {
+    return optMasterAddresses;
+  }
+
+  public String getOptTableName() {
+    return optTableName;
+  }
+
+  public String getOptSchema() {
+    return optSchema;
+  }
+
+  @Nullable
+  public String getOptColumnProjection() {
+    return optColumnProjection;
+  }
+
+  @Nullable
+  public String getOptOperationTimeout() {
+    return optOperationTimeout;
+  }
 
   /**
    * Validates the configuration fields.
    *
    * @throws IllegalArgumentException throw when there is issue with the configuration.
    */
-  public void validate() throws IllegalArgumentException {
-    if (!containsMacro("master")) {
-      Preconditions.checkArgument(!Strings.isNullOrEmpty(optMasterAddresses),
-                                  "Kudu Master Server address list is empty.");
+  public void validate(FailureCollector failureCollector) throws IllegalArgumentException {
+    if (!containsMacro(MASTER)){
+      if (Strings.isNullOrEmpty(optMasterAddresses)) {
```

Comment:
remove these checks. Add or not add @nullable instead to control this.

Reply:
Done